### ISSUE TITLE
regex-base-0.94.0.0 broken by ghc-9.0.1 alpha release

### ISSUE THREAD

**Issue by peti** [haskell-hvr/regex-base#7]
The new white-space rules for prefix operators break the build:

~~~
src/Text/Regex/Base/Context.hs:303:29: error:
    Bang pattern in expression context: !0
    Did you mean to add a space after the '!'?
    |
303 |   match r s = maybe (-1,0) (!0) (matchOnce r s)
    |                             ^^

src/Text/Regex/Base/Context.hs:304:43: error:
    Bang pattern in expression context: !0
    Did you mean to add a space after the '!'?
    |
304 |   matchM r s = maybe regexFailed (return.(!0)) (matchOnce r s)
    |                                           ^^
~~~

**Comment by peti** [comment]
This patch fixes the issue: https://gitlab.haskell.org/ghc/head.hackage/-/blob/master/patches/regex-base-0.94.0.0.patch

**Comment by hvr** [comment]
Hello Peter,

> The new white-space rules for prefix operators break the build:
>
> ~~~
> src/Text/Regex/Base/Context.hs:303:29: error:
>     Bang pattern in expression context: !0
>     Did you mean to add a space after the '!'?
>     |
> 303 |   match r s = maybe (-1,0) (!0) (matchOnce r s)
>     |                             ^^
>

I'm afraid I don't understand the problen here;

`regex-base-0.94.0.0` is declared as Haskell2010 in the .cabal file and
the module in question merely enables the extensions

{-# LANGUAGE MultiParamTypeClasses, FlexibleInstances, FlexibleContexts, TypeSynonymInstances #-}

I don't get why GHC thinks that the BangPatterns extensions has any
relevance here; unless I'm mistaken, under Haskell98 and Haskell2010, an
expression such as

   (!0)

is supposed to be an operator section, also it's appearing on the RHS of
a definition, so I'm really at loss here why this would need changing at
this point. I don't have time right now to review the grammar rules from
the Haskell2010 report, but I'd be really surprised if I'd been wrongly
assuming such an expression to be valid Haskell2010 for almost 2
decades... I think it's more likely this is a regression in the upcoming
GHC 9?

**Comment by phadej** [comment]
This is not a regression. This is a consequence of https://github.com/ghc-proposals/ghc-proposals/blob/master/proposals/0229-whitespace-bang-patterns.rst

EDIT: it is a regression, but deliberate.

**Comment by phadej** [comment]
See https://github.com/haskell-hvr/regex-base/pull/3/files

**Comment by galenhuntington** [comment]
Any chance we can get one of #2, #3, or #6 merged?  That change seems completely harmless, and many packages transitively depend on this one and so will fail to build on GHC 9.

**Comment by phadej** [comment]
@galenhuntington GHC-9.0 is not released, and it is IMHO somewhat impolite to pressure maintainers to merge (and release) patches. Note, this issue was opened in late October, and GHC-9.0-rc1 was released just last week.

I'm sure Herbert will eventually get to this issue, until then, there is patch in https://gitlab.haskell.org/ghc/head.hackage

**Comment by asr** [comment]
Blocking https://github.com/agda/agda/issues/4955.

**Comment by andreasabel** [comment]
I am working on this.

**Comment by andreasabel** [comment]
Fixed by release of 0.94.0.1, see https://hackage.haskell.org/package/regex-base-0.94.0.1.
